### PR TITLE
feat(query): Support select * exclude [column_name | (col_name, col_name,...)]

### LINKED ISSUE #8964: Feature: selecttarget QualifiedName support EXCLUDE(col)
**Summary**

select a.* [exclude | except] col1 from xx;

select * [exclude | except] (col1, col2, ...)

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Now support query like this:

```sql
select * exclude col1 from t;
select * exclude (col1, col2) from t;
```

These solution will return err:

1. if col_name in except is duplicate. e.g. `except (id, id)`
2. col_name is not exists in table. e.g. `t.* except (id) from t -- table t does not have id column`
3. col_name is ambiguous. e.g. `except (id) from (select col1 as id, col2 as id)`

Closes #8964

### CODE REVIEW

**Comment by sundy-li** [comment]
Keyword `EXCEPT` should not be allowed to use here. It's ambiguous with  union expect query,  let's just use EXCLUDE.

File: src/query/sql/src/planner/binder/project.rs [hunk 1]
```diff
@@ -150,27 +153,150 @@ impl<'a> Binder {
...
+                    exclude,
+                } => {
                     // Handle qualified name as select target
+                    let mut exclude_cols: HashSet<String> = HashSet::new();
+                    if let Some(exclude) = exclude {
+                        match exclude {
+                            ExcludeCol::Col(col) => {
+                                exclude_cols.insert(col.name.clone());
+                            }
+                            ExcludeCol::Cols(cols) => {
```
Comment (xudong963) [inline src/query/sql/src/planner/binder/project.rs:167]: Why `ExcludeCol::Col` and `ExcludeCol::Cols` are concomitant?  

When `Cols` has only one column, isn't it `Col`?

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs:167]: I think we need judge `* exclude (id, id) ` the col is duplicate.

File: src/query/ast/src/ast/query.rs [hunk 2]
```diff
@@ -119,9 +119,31 @@ pub enum SelectTarget<'a> {
         alias: Option<Identifier<'a>>,
     },
 
-    // Qualified name, e.g. `SELECT t.a, t.* FROM t`.
+    // Qualified name, e.g. `SELECT t.a, t.* exclude t.a FROM t`.
     // For simplicity, wildcard is involved.
-    QualifiedName(QualifiedName<'a>),
+    QualifiedName {
+        qualified: QualifiedName<'a>,
+        exclude: Option<ExcludeCol<'a>>,
```
Comment (b41sh) [inline src/query/ast/src/ast/query.rs:126]: It seems too complicated, can we use `exclude: Option<Vec<Identifier<'a>>>`?

Comment (TCeason) [reply src/query/ast/src/ast/query.rs:126]: I think it's ok.

File: src/query/sql/src/planner/binder/project.rs [hunk 3]
```diff
@@ -150,27 +152,143 @@ impl<'a> Binder {
...
+                    // Pre-check exclude_col is legal
+                    let precheck_exclude_cols = |input_context: &BindContext,
+                                                 exclude_cols: &HashSet<String>,
+                                                 db_name: Option<&Identifier>,
+                                                 table_name: Option<&Identifier>|
+                     -> Result<()> {
+                        let all_columns_bind = input_context.all_column_bindings();
+                        let mut qualified_cols_name: HashMap<String, u8> = HashMap::new();
+                        for i in all_columns_bind {
+                            if let (None, None) = (db_name, table_name) {
```
Comment (xudong963) [inline src/query/sql/src/planner/binder/project.rs:180]: I recommend to use `match (db_name, table_name) {}`

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs:180]: I use this match first. But I think this arm is unnecessary.

```
match  (db_name, table_name) {
   (Some(db_name), None ) => {}
}
```

So I use `if let`.

Comment (Xuanwo) [reply src/query/sql/src/planner/binder/project.rs:180]: > But I think this arm is unnecessary.

Although there is an unnecessary arm, they are more readable than three `if let` cases.

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs:180]: > > But I think this arm is unnecessary.
> 
> Although there is an unnecessary arm, they are more readable than three `if let` cases.

Agreed, it's better in terms of readability. Let's change it.

File: src/query/sql/src/planner/binder/project.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/sql/src/planner/binder/project.rs]: Seems  `ExcludeCol::Cols(cols)` can cover `ExcludeCol::Col(col) `, so I recommend to keep `Cols`.

An extra judge for single item vec is trivial, but the code could be cleaner.

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs]: > Seems `ExcludeCol::Cols(cols)` can cover `ExcludeCol::Col(col) `, so I recommend to keep `Cols`.
> 
> An extra judge for single item vec is trivial, but the code could be cleaner.

Now it already change to

```rust
QualifiedName {
  qualified: QualifiedName<'a>,
  exclude: Option<Vec<Identifier<'a>>>,
},
```

So we can drop struct ExcludeCol

File: src/query/sql/src/planner/binder/project.rs (hunk 3 above)
Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs]: > I recommend to use `match (db_name, table_name) {}`

Done

File: src/query/sql/src/planner/binder/project.rs [hunk 4]
```diff
@@ -150,27 +152,151 @@ impl<'a> Binder {
...
+                    // Pre-check exclude_col is legal
+                    let precheck_exclude_cols = |input_context: &BindContext,
+                                                 exclude_cols: &HashSet<String>,
+                                                 db_name: Option<&Identifier>,
+                                                 table_name: Option<&Identifier>|
+                     -> Result<()> {
+                        let all_columns_bind = input_context.all_column_bindings();
+                        let mut qualified_cols_name: HashMap<String, u8> = HashMap::new();
+                        for i in all_columns_bind {
+                            match (db_name, table_name) {
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/project.rs:180]: better to use `match_column_binding` to match columns.

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs]: Can't use `match_column_binding`. But it gives me inspiration. Now we use match wrap for loop and extract a function named `fill_qualified_cols`

**Review by xudong963 — APPROVED** [review]
The `normalize_select_list` method is toooo tedious :/

Better to split it (Doing this in the next PR also makes sense to me)

**Comment by TCeason** [comment]
> The `normalize_select_list` method is toooo tedious :/
> 
> Better to split it (Doing this in the next PR also makes sense to me)

Try to merge * and table.* in the newest commit. Please review it.

File: src/query/sql/src/planner/binder/project.rs [hunk 5]
```diff
@@ -290,3 +308,79 @@ impl<'a> Binder {
...
+) -> Result<()> {
+    let all_columns_bind = input_context.all_column_bindings();
+    let mut qualified_cols_name: HashMap<String, u8> = HashMap::new();
+
+    fn fill_qualified_cols(
+        qualified_cols_name: &mut HashMap<String, u8>,
+        column_bind: &ColumnBinding,
+    ) {
+        let col_name = column_bind.column_name.clone();
+        if qualified_cols_name.contains_key(col_name.as_str()) {
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/project.rs:327]: we can return `ambiguous column name '{exclude_col}'` here, without contiune check.

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs:327]: > we can return `ambiguous column name '{exclude_col}'` here, without contiune check.

Done

File: src/query/sql/src/planner/binder/project.rs [hunk 6]
```diff
@@ -290,3 +308,75 @@ impl<'a> Binder {
...
+    input_context: &BindContext,
+    exclude_cols: &HashSet<String>,
+    db_name: Option<&Identifier>,
+    table_name: Option<&Identifier>,
+) -> Result<()> {
+    let all_columns_bind = input_context.all_column_bindings();
+    let mut qualified_cols_name: HashMap<String, u8> = HashMap::new();
+
+    fn fill_qualified_cols(
+        qualified_cols_name: &mut HashMap<String, u8>,
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/project.rs:323]: use `HashSet`

Comment (b41sh) [inline src/query/sql/src/planner/binder/project.rs:323]: use `HashSet`

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs:323]: Truly, the value is no used. HashSet is better

Comment (TCeason) [reply src/query/sql/src/planner/binder/project.rs]: > use `HashSet`

Fixed

File: src/query/ast/src/parser/query.rs [hunk 7]
```diff
@@ -83,21 +83,48 @@ pub fn with(i: Input) -> IResult<With> {
...
+pub fn exclude_col(i: Input) -> IResult<Vec<Identifier>> {
+    let var = map(
+        rule! {
+            #ident
+        },
+        |col| vec![col],
+    );
+    let vars = map(
+        rule! {
+            "(" ~ ^#comma_separated_list1(ident) ~ ")"
```
Comment (andylokandy) [inline src/query/ast/src/parser/query.rs:95]:
```suggestion
            "(" ~ ^#comma_separated_list1(ident) ~ ^")"
```

File: src/query/ast/src/parser/query.rs [hunk 8]
```diff
@@ -83,21 +83,48 @@ pub fn with(i: Input) -> IResult<With> {
...
+                exclude: opt_exclude.map(|(_, exclude)| exclude),
+            },
+            Some((fst, _, None)) => SelectTarget::QualifiedName {
+                qualified: vec![Indirection::Identifier(fst), Indirection::Star],
+                exclude: opt_exclude.map(|(_, exclude)| exclude),
+            },
+            None => SelectTarget::QualifiedName {
+                qualified: vec![Indirection::Star],
+                exclude: opt_exclude.map(|(_, exclude)| exclude),
+            },
```
Comment (andylokandy) [inline src/query/ast/src/parser/query.rs:127]:
```suggestion
        |(res, _, opt_exclude)| {
            let exclude = opt_exclude.map(|(_, exclude)| exclude);
            match res {
                Some((fst, _, Some((snd, _)))) => SelectTarget::QualifiedName {
                    qualified: vec![
                        Indirection::Identifier(fst),
                        Indirection::Identifier(snd),
                        Indirection::Star,
                   ],
                    exclude,
                },
                Some((fst, _, None)) => SelectTarget::QualifiedName {
                    qualified: vec![Indirection::Identifier(fst), Indirection::Star],
                    exclude,
                },
                None => SelectTarget::QualifiedName {
                    qualified: vec![Indirection::Star],
                    exclude,
                },
            }
```

File: src/query/ast/src/parser/query.rs (hunk 7 above)
Comment (TCeason) [reply src/query/ast/src/parser/query.rs]: Done

File: src/query/ast/src/parser/query.rs (hunk 8 above)
Comment (TCeason) [reply src/query/ast/src/parser/query.rs]: Done

**Review by b41sh — APPROVED** [review]

**Comment by TCeason** [comment]
cc @soyeric128 Now we can add doc in https://databend.rs/doc/sql-commands/query-syntax/dml-select

**Pull request merged**